src/lib/statistics/statistics_objects/abstract_histogram.cpp

```diff
@@ -0,0 +1,1035 @@
+    } else {
+      // TODO(tim): turn into compile-time error
+      Fail("Not supported for strings.");
+    }
+  }
+
+  for (const auto& edge_pair : additional_bin_edges) {
+    if constexpr (std::is_arithmetic_v<
+                      T>) {  // NOLINT clang-tidy is crazy and sees a "potentially unintended semicolon" here...
+      split_set.insert(std::make_pair(_domain.previous_value(edge_pair.first), edge_pair.first));
+      split_set.insert(std::make_pair(edge_pair.second, _domain.next_value(edge_pair.second)));
+    } else {
+      // TODO(tim): turn into compile-time error
+      Fail("Not supported for strings.");
+    }
+  }
+
+  /**
+   * Split pairs into single values.
+   * These are the new bin edges, and we have to sort them.
+   */
+  std::vector<T> all_edges(split_set.size() * 2);
+  size_t edge_idx = 0;
+  for (const auto& split_pair : split_set) {
+    all_edges[edge_idx] = split_pair.first;
+    all_edges[edge_idx + 1] = split_pair.second;
+    edge_idx += 2;
+  }
+
+  std::sort(all_edges.begin(), all_edges.end());
+
+  /**
+   * Remove the first and last value.
+   * These are the values introduced in the first step for the first and last split.
+   * The lower edge of bin 0 basically added the upper edge of bin -1,
+   * and the upper edge of the last bin basically added the lower edge of the bin after the last bin.
+   * Both values are obviously not needed.
+   */
+  all_edges.erase(all_edges.begin());
+  all_edges.pop_back();
+
+  // We do not resize the vectors because we might not need all the slots because bins can be empty.
+  const auto new_bin_count = all_edges.size() / 2;
+
+  GenericHistogramBuilder<T> builder{new_bin_count, _domain};
+
+  /**
+   * Create new bins.
+   * Bin edges are defined by consecutive values in pairs of two.
+   */
+  for (auto bin_id = BinID{0}; bin_id < new_bin_count; bin_id++) {
+    const auto bin_min = all_edges[bin_id * 2];
+    const auto bin_max = all_edges[bin_id * 2 + 1];
+
+    const auto estimate =
+        estimate_cardinality_and_distinct_count(PredicateCondition::BetweenInclusive, bin_min, bin_max);
+    // Skip empty bins
+    if (estimate.cardinality == Cardinality{0}) {
+      continue;
+    }
+
+    builder.add_bin(bin_min, bin_max, static_cast<HistogramCountType>(estimate.cardinality),
+                    static_cast<HistogramCountType>(estimate.distinct_count));
+  }
+
+  return builder.build();
+}
+
+template <typename T>
+std::vector<std::pair<T, T>> AbstractHistogram<T>::bin_bounds() const {
+  std::vector<std::pair<T, T>> bin_edges(bin_count());
+
+  for (auto bin_id = BinID{0}; bin_id < bin_edges.size(); bin_id++) {
+    bin_edges[bin_id] = std::make_pair(bin_minimum(bin_id), bin_maximum(bin_id));
+  }
+
+  return bin_edges;
+}
+
+template <typename T>
+void AbstractHistogram<T>::_assert_bin_validity() {
+  for (BinID bin_id{0}; bin_id < bin_count(); ++bin_id) {
+    Assert(bin_minimum(bin_id) <= bin_maximum(bin_id), "Cannot have overlapping bins.");
```

Comment:
You don't check for overlapping here, but for the maximum of a bin being bigger than its own (!) minimum

Reply:
Wrong message, is now "Bin minimum must be <= bin maximum"